---
title: window.onerror
order: 2
---

## 这是什么玩意儿

`window.onerror` 是浏览器自带的 api，是用来捕捉 js 代码中所暴露的错误信息，`window.onerror` 一共有五个参数，分别是：

1. `msg`: 错误发生的异常信息
2. `url`: 错误发生的脚本 URL
3. `lineNum`: 错误发生的行号
4. `colNum`: 错误发生的列号
5. `err`: 错误发生的 Error 对象
   > 注意： **`window.onerror` 只能捕捉 throw 出来的错误，无法捕捉 catch 的错误，比如 Promise 的 catch 错误是无法拿到的。**

## 怎么使用

`window.onerror` 使用起来很简单，只需要在最外层写一个监听函数就行了，举个栗子：

```js
window.onerror = function (msg, url, lineNum, colNum, err) {
  console.log(`错误发生的异常信息: ${msg}`);
  console.log(`错误发生的脚本URL: ${url}`);
  console.log(`错误发生的行号: ${lineNum}`);
  console.log(`错误发生的列号: ${colNum}`);
  console.log(`错误发生的Error对象: ${err}`);
};

throw new Error('这是一个错误');
/**
错误发生的异常信息: Uncaught Error: 这是一个错误
错误发生的脚本URL: file:///Users/relx/Desktop/study/index.html
错误发生的行号: 8
错误发生的列号: 7
错误发生的Error对象: Error: 这是一个错误
*/
```

> `window.onerror` 可以 `return true` 让错误信息不输出到控制台。  
> `window.onerror` 监听函数里面的语法错误或者其他错误也无法被监听，因为监听函数都无法执行，当然无法执行监听了。  
> `window.onerror` 只能监听在监听函数执行之后的代码，不能监听之前的代码，所以一般需要将`window.onerror`绑定事件放在代码最前面执行。

## 跨域问题

`window.onerror` 虽然很牛逼，但是也是有一定的问题，如果在真实开发中，我们总会引用很多外部的库，很多第三方 js 文件，这时候如果使用 `window.onerror` 去监听的话，会发现第三方 js 文件里面的一些错误信息会成为下面的样子:
|错误信息|错误 url|错误行号|错误列号|错误对象|
|:----:|:----:|:----:|:----:|:----:|
|Script error|空|0|0|null|

会发现，拿到的错误全部都是没有用的参数，这是为什么呢？这是因为 `浏览器的同源策略` 问题。

> 在浏览器中，当加载自不同域（协议、域名、端口三者任一不同）的脚本中发生语法错误时，为避免信息泄露，语法错误的细节将不会报告，而代之简单的"Script error."。
> 那么如何解决这个问题呢？

## 解决 onerror 跨域问题

首先我们需要给我们的 script 标签添加 `crossorigin` 属性，其次需要在我们的引用的 js 文件的响应头里面添加 `Access-Control-Allow-Origin: *` 字段，这样子就可以拿到正确的错误信息了。

不过 `Access-Control-Allow-Origin: *` 响应头是需要服务端去设置的，所以一般情况下是需要第三方库主动添加响应头才可以，孤掌难鸣嘛。
